cudax/include/cuda/experimental/__container/vector.h

```diff
@@ -0,0 +1,426 @@
+  }
+
+  //! @brief Returns pointer to the start of the vector
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr const_reverse_iterator crbegin() const noexcept
+  {
+    return const_reverse_iterator{end()};
+  }
+
+  //! @brief Returns pointer to the end of the vector
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr reverse_iterator rend() noexcept
+  {
+    return reverse_iterator{begin()};
+  }
+
+  //! @brief Returns pointer to the start of the vector
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr const_reverse_iterator rend() const noexcept
+  {
+    return const_reverse_iterator{begin()};
+  }
+
+  //! @brief Returns pointer to the start of the vector
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr const_reverse_iterator crend() const noexcept
+  {
+    return const_reverse_iterator{begin()};
+  }
+
+  //! @brief Returns pointer to the start of the vector
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr pointer data() noexcept
+  {
+    return __buf_.data();
+  }
+
+  //! @brief Returns pointer to the start of the vector
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr const_pointer data() const noexcept
+  {
+    return __buf_.data();
+  }
+  //! @}
+
+  //! @addtogroup capacity
+  //! @{
+  //! @brief Returns the size of the vector
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr size_t size() const noexcept
+  {
+    return __size_;
+  }
+
+  //! @brief Returns true if the vector is empty
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr bool empty() const noexcept
+  {
+    return __size_ == 0;
+  }
+
+  //! @brief Returns the capacity of the vector
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr size_t capacity() const noexcept
+  {
+    return __buf_.size();
+  }
+
+  //! @rst
+  //! Returns the :ref:`resource_ref <libcudacxx-extended-api-memory-resources-resource-ref>` used to allocate
+  //! the buffer
+  //! @endrst
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE _CUDA_VMR::resource_ref<_Properties...> resource() const noexcept
+  {
+    return __mr_;
+  }
+  //! @}
+
+  //! @addtogroup access
+  //! @{
+  //! @brief Returns a reference to the \p __n 'th element of the vector
+  //! @param __n The index of the element we want to access
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr reference operator[](const size_t __n) noexcept
+  {
+    return *(__buf_.data() + __n);
+  }
+
+  //! @brief Returns a reference to the \p __n 'th element of the vector
+  //! @param __n The index of the element we want to access
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr const_reference operator[](const size_t __n) const noexcept
+  {
+    return *(__buf_.data() + __n);
+  }
+
+  //! @brief Returns a reference to the first element of the vector
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr reference first() noexcept
+  {
+    return *__buf_.data();
+  }
+
+  //! @brief Returns a reference to the first element of the vector
+  _CCCL_NODISCARD _CCCL_HOST_DEVICE constexpr const_reference first() const noexcept
+  {
+    return *__buf_.data();
```

Comment:
These accessors can't be implemented the same for host and device except if the memory is both host- and device-accessible!  Why don't these use the execution space functionality of the `heterogeneous_iterator`?

Reply:
Yeah this is WIP, I have some work I need to push

Reply:
Sorry, I missed that this is in draft.

Reply:
I did not push, so there was nothing to see